### PR TITLE
Make dropout configurable per-layer, fixes #372

### LINKED ISSUE #372: Make dropout more configurable
Brief: It would be advantageous if we had greater control over the dropout that gets applied to each layer in our model.

After presenting on the paper 'Swapout: Learning an ensemble of deep architectures'[0], I performed some experimentation to see the results of differing levels of dropout, ResNet, swapout, etc. on the DeepSpeech model. An important part of the positive results in the swapout paper was applying a linearly-increasing amount of dropout depending on the depth of the layers. This makes intuitive sense, as any stochasticity applied at the top of the model will have an exponentially increasing effect as it travels deeper into the tree.

Performing tests with a limited sample of the TED data set, it would appear that a dropout of lin(0.1,0.4) (i.e. dropouts increasing linearly from 0.1 to 0.4 on the feed-forward layers of the network) outperforms dropout of 0.3, when measuring validation loss and WER. Further testing also showed that applying dropout of 0.2 on the RNN layers of the network also had positive effects on validation loss and WER.

While further experimentation is needed to determine optimal values, it would be advantageous to have the parameters to configure and test these values in our implementation.

[0] https://docs.google.com/presentation/d/1dyl7F_sW0YLa3HchHVwugdBlvFzwbWRSiCRjD-iX2Yw/edit?usp=sharing

### PR DESCRIPTION
**Pull request by Cwiiis**

### CODE REVIEW

**Comment by kdavis-mozilla** [comment]
Is there an issue associated with this PR?

**Comment by Cwiiis** [comment]
There's no issue associated with this PR, shall I open one?

File: DeepSpeech.py [hunk 1]
```diff
@@ -231,27 +246,35 @@ def BiRNN(batch_x, seq_length, dropout):
...
+    layer_3 = tf.nn.dropout(layer_3, (1.0 - dropout[2]))
 
     # Now we create the forward and backward LSTM units.
     # Both of which have inputs of length `n_cell_dim` and bias `1.0` for the forget gate of the LSTM.
 
     # Forward direction cell:
     lstm_fw_cell = tf.nn.rnn_cell.BasicLSTMCell(n_cell_dim, forget_bias=1.0, state_is_tuple=True)
+    lstm_fw_cell = tf.nn.rnn_cell.DropoutWrapper(lstm_fw_cell,
+                                                 input_keep_prob=1.0 - dropout[3],
+                                                 output_keep_prob=1.0 - dropout[3],
```
Comment (kdavis-mozilla) [inline DeepSpeech.py]: Why are input and output dropout the same?

File: DeepSpeech.py [hunk 2]
```diff
@@ -231,27 +246,35 @@ def BiRNN(batch_x, seq_length, dropout):
...
+                                                 seed=random_seed)
     # Backward direction cell:
     lstm_bw_cell = tf.nn.rnn_cell.BasicLSTMCell(n_cell_dim, forget_bias=1.0, state_is_tuple=True)
+    lstm_bw_cell = tf.nn.rnn_cell.DropoutWrapper(lstm_bw_cell,
+                                                 input_keep_prob=1.0 - dropout[4],
+                                                 output_keep_prob=1.0 - dropout[4],
```
Comment (kdavis-mozilla) [inline DeepSpeech.py]: Why are input and output dropout the same?

**Comment by kdavis-mozilla** [comment]
@Cwiiis Could you open an issue, just so we have context. Maybe even mention the Swapout paper.

**Comment by Cwiiis** [comment]
I've filed issue #372 to track this. As to why input/output are the same, no good reason - the examples I saw of others applying dropout to RNN layers, they did it this way. I don't know how applying different levels to input/output would affect results - intuitively, I would say it wouldn't in a way that couldn't be replicated by just picking a different value and applying it to both input and output (i.e. input_keep=0.8, output_keep=1.0 may be roughly equivalent to input_keep=output_keep=0.9), but that may be based on a misunderstanding of how it works.
